v5.12 changelog (#2706)
mattermost/docs: source/administration/changelog.md

### config.json

Multiple setting options were added to `config.json`. Below is a list of the additions and their default values on install. The settings can be modified in `config.json`, or the System Console when available.

#### Changes to Team Edition and Enterprise Edition:

 - Under ``"PluginSettings":`` in ``config.json``:
   - Added ``"EnableHealthCheck": true``, to ensure all plugins are periodically monitored, and restarted or deactivated based on their health status.
 - Under ``"NotificationLogSettings":`` in ``config.json``:
   - Added ``"EnableConsole": true``, ``"ConsoleLevel": "DEBUG"``, ``"ConsoleJson": true``, ``"EnableFile": true``, ``"FileLevel": "INFO"``, ``"FileJson": true``, and ``"FileLocation": ""``, to implement a structured logger to keep track of push notifications.
 - Under ``"ServiceSettings":`` in ``config.json``:
   - Added ``"EnableBotAccountCreation": false`` to enable bot account creation.
   - Added ``"DisableBotsWhenOwnerIsDeactivated": true`` to disable bots automatically when the owner is deactivated.
   - Added ``"TrustedProxyIPHeader": []``, to explicitly define which IP headers are trusted.

### Database Changes

>  - 

Reviewer: ```
`SchemeGuest` column added to the `TeamMembers` table.
`SchemeGuest` column added to the `ChannelMembers` table.
`DefaultTeamGuestRole` column added to the `Schemes` table, and set to an empty string.
`DefaultChannelGuestRole` column added to the `Schemes` table, and set to an empty string.
```

(from https://github.com/mattermost/mattermost-server/blob/master/store/sqlstore/upgrade.go#L683)

@jespino may be able to confirm if the above is correct

Reviewer 2: Yes, the @jasonblais list of changes is correct, I added that fields as part of the guest account preparation changes.

### API Changes

#### RESTful API v4 Changes
 - Updated API to use gziphandler wrapper if server is configured to use gzip. This ensures that the Mattermost server can respond to REST API requests with compressed data (via gzip) to reduce the amount of bandwidth used.
 - LDAP Group Sync:

>     - Added API endpoints ``getGroupsByChannel`` and ``GetGroupsByTeam`` to retrieve groups by team and by channel.

Reviewer: Propose asking Martin to confirm endpoints for LDAP group sync

Author: @skiplist ^

Reviewer 3: Also new APIs for...

* List/search groups
* Regenerate team invite ids
* Invalidate team email invites
* `/notifications/ack` (something to do with push notifications)

Let me know if you need more detail.

    - Added ``group_constrained`` API to both ``/users`` and ``/users/search`` endpoints to be able to limit users listed to those allowed by group-constraints.
    
#### Plugin API v4 Changes
 - Added ``KVCompareAndSet(key string, old []byte, new []byte)`` to Plugin API to add support for transactional semantics with KV Store in plugin framework.

### Websocket Event Changes

>  - 

Reviewer: Maybe check with SET?

Author: @enahum / @jespino Were there any Websocket changes in 5.12?

Reviewer 4: Not that I’m aware of

 
#### LDAP Group Sync Removals
 - Users can manage the membership of private teams and channels with groups, eliminating the need to individually add and remove members. Users in the groups will be automatically removed from the team or channel when removed from an associated group.
 
#### User/Channel Search & AutoComplete in ElasticSearch
 - Added a new flag in the **System console > Elasticsearch** to enable the usage of Elasticsearch for autocompletion queries. When this flag is enabled, Elasticsearch will use its indexed data for the autocompletions.

### Improvements

#### User Interface (UI)
 - Added an option to add a user to a channel from the profile pop-over.
 - Removed ``@`` for full name display in push notifications.
 
#### Plugins
 - Added support for markdown in plugin system console help text fields.
 - Added support for plugins to override ephemeral posts.

#### Localization
 - Promoted Polish language to "official".

#### Command Line Interface (CLI)
 - Added a ``command modify`` CLI command to be able to modify slash commands.

>  - Added support for converting user accounts to bot accounts through the CLI.

Reviewer 5: I found a similar command here [mattermost user convert --bot github](https://github.com/mattermost/mattermost-plugin-github/blob/master/README.md#step-3-configure-plugin-in-mattermost), is this the new command for user group?

>  - Added support for converting user accounts to bot accounts through the CLI.

Reviewer: `mattermost user convert --bot` is indeed the new CLI command to convert user accounts to bots 👍

 - Implemented a new command ``MigrateConfig`` for migrating configuration to and from the database.
 - For LDAP Groups, added the ability to switch a team or channel to be group-constrained via the CLI.
 
#### Administration
 - Added support for running two Mattermost instances on the same domain using subpaths.

>  - Added support for importing threads from Slack.

Reviewer: what about the corporate Slack import?

Author: The PR hasn't been merged yet. Should I add it already?

Reviewer: Ah, didn't realize it wasn't merged yet. I guess it's gonna be too late for 5.12. Are we doing a custom build for the customer instead?

 - Fixed an issue where replies to the parent post were not left-aligned.
 - Fixed an issue where the timezone picker dropdown closed when trying to drag the scrollbar.
 - Fixed an issue where the ``ExperimentalPrimaryTeam`` config.json setting no longer hid the "Leave Team" option.

### config.json

Multiple setting options were added to `config.json`. Below is a list of the additions and their default values on install. The settings can be modified in `config.json`, or the System Console when available.

#### Changes to Team Edition and Enterprise Edition:

 - Under ``"PluginSettings":`` in ``config.json``:
   - Added ``"EnableHealthCheck": true``, to ensure all plugins are periodically monitored, and restarted or deactivated based on their health status.
 - Under ``"NotificationLogSettings":`` in ``config.json``:
   - Added ``"EnableConsole": true``, ``"ConsoleLevel": "DEBUG"``, ``"ConsoleJson": true``, ``"EnableFile": true``, ``"FileLevel": "INFO"``, ``"FileJson": true``, and ``"FileLocation": ""``, to implement a structured logger to keep track of push notifications.
 - Under ``"ServiceSettings":`` in ``config.json``:

>    - Added ``"EnableBotAccountCreation": false`` and ``"DisableBotsWhenOwnerIsDeactivated": true``, to add the ability to enable bot accounts.

Reviewer: ```suggestion
   - Added ``"EnableBotAccountCreation": false`` to enable bot account creation.
   - Added ``"DisableBotsWhenOwnerIsDeactivated": true`` to disable bots automatically when the owner is deactivated.
```

   - Added ``"TrustedProxyIPHeader": []``, to explicitly define which IP headers are trusted.

### Database Changes
 - 

> ### API Changes

Reviewer: 1) Propose breaking to "RESTful API Changes" and "Plugin API" changes (for plugins there is only one, the last one).

![image](https://user-images.githubusercontent.com/13119842/58820456-aa7dbd80-8600-11e9-82de-faae02c3d453.png)

This helps make it more clear which APIs the changes refer to.

2) Propose formatting the API changes consistently, similar to the screenshot above. Let me know if I can help.

 - Updated API to use gziphandler wrapper if server is configured to use gzip. This ensures that the Mattermost server can respond to REST API requests with compressed data (via gzip) to reduce the amount of bandwidth used.
 - Added an API command to create and delete a team and channel.
 - LDAP Group Sync:
    - Added API endpoints ``getGroupsByChannel`` and ``GetGroupsByTeam`` to retrieve groups by team and by channel.
    - Added ``group_constrained`` API to both ``/users`` and ``/users/search`` endpoints to be able to limit users listed to those allowed by group-constraints.
 - Added ``KVCompareAndSet(key string, old []byte, new []byte)`` to Plugin API to add support for transactional semantics with KV Store in plugin framework.

### Websocket Event Changes
 - 

### Known Issues

>  - Buttons inside ephemeral posts are not clickable / functional on the mobile app.

Reviewer: ```suggestion
 - Buttons inside ephemeral messages are not clickable / functional on the mobile app.
```

>  - On a server using a subpath, the URL opens a blank page if the system admin changes the Site URL in the System Console UI. The system admin should restart the server to fix it.

Reviewer: ```suggestion
 - On a server using a subpath, the URL opens a blank page if the System Admin changes the Site URL in the System Console UI. To fix, the System Admin should restart the server.
```

 
#### Bot Accounts
 - Users no longer have to rely on creating fake user accounts to act as bots for integrations. Instead, create a real bot account and use it to generate bot access tokens to interact with users and complete tasks.
 - Users can can also use these bots to post to any channel in the system, whether it’s a private team, private channel or a direct message channel.
 - For Enterprise deployments, bot accounts no longer count as an active user towards licensing subscriptions.
 
#### Jira Plugin 2.0
 - Enhanced existing plugin for a deep two-way integration between Jira and Mattermost.
 - Send notifications for Jira issue creation, issue updates and comments to Mattermost channels.
 - Users can also take quick actions in Mattermost, including creating Jira issues, attaching Mattermost messages to Jira issues, and transitioning issues via slash commands.
 - For a full feature set for 2.0, see https://github.com/mattermost/mattermost-plugin-jira#jira-20-features.

> 

Reviewer: As another highlight, we can include the list of new pre-packaged plugins we're shipping in 5.12. (Would be good to add to release announce as well)

```
#### Pre-packaged Plugins
 - New pre-packaged plugins bundled with this Mattermost release include:
   - [GitHub plugin](https://github.com/mattermost/mattermost-plugin-github) for notifications, reminders and slash commands to stay up-to-date on issues and pull requests. Supports GitHub SaaS and Enterprise versions.
   - [Autolink plugin](https://github.com/mattermost/mattermost-plugin-autolink) to automatically hyperlink text, such as adding links to your issue tracker when someone posts an issue key or number.
   - [Custom Attributes plugin](https://github.com/mattermost/mattermost-plugin-custom-attributes) to add custom attributes in the user profile popover.
   - [Welcome Bot plugin](https://github.com/mattermost/mattermost-plugin-welcomebot) to improve onboarding and HR processes by adding a Welcome Bot that helps add new team members to channels.
   - [Amazon SNS CloudWatch plugin](https://github.com/mattermost/mattermost-plugin-aws-sns) to send alert notifications from [Amazon AWS CloudWatch](https://aws.amazon.com/cloudwatch/) to Mattermost channels via AWS SNS.

#### Changes to Team Edition and Enterprise Edition:

 - Under ``"PluginSettings":`` in ``config.json``:
   - Added ``"EnableHealthCheck": true``, to ensure all plugins are periodically monitored, and restarted or deactivated based on their health status.
 - Under ``"NotificationLogSettings":`` in ``config.json``:
   - Added ``"EnableConsole": true``, ``"ConsoleLevel": "DEBUG"``, ``"ConsoleJson": true``, ``"EnableFile": true``, ``"FileLevel": "INFO"``, ``"FileJson": true``, and ``"FileLocation": ""``, to implement a structured logger to keep track of push notifications.
 - Under ``"ServiceSettings":`` in ``config.json``:
   - Added ``"EnableBotAccountCreation": false`` to enable bot account creation.
   - Added ``"DisableBotsWhenOwnerIsDeactivated": true`` to disable bots automatically when the owner is deactivated.
   - Added ``"TrustedProxyIPHeader": []``, to explicitly define which IP headers are trusted.

### Database Changes
 - 

### API Changes
 - Updated API to use gziphandler wrapper if server is configured to use gzip. This ensures that the Mattermost server can respond to REST API requests with compressed data (via gzip) to reduce the amount of bandwidth used.

>  - Added an API command to create and delete a team and channel.

Reviewer: This is actually for Cypress end-to-end testing only, not for interacting with Mattermost https://github.com/mattermost/mattermost-webapp/pull/2601#issuecomment-498448815
